### ISSUE TITLE
ENH: `stats.gaussian_kde`: avoid `ValueError` unless necessary when `n_dimensions > n_data`

### ISSUE THREAD

**Issue by csbrown** [scipy/scipy#21685]
### Describe your issue.

`gaussian_kde` throws a `ValueError` when the number of dimensions is greater than the number of data points.  This only affects certain algorithms: a KDE is still perfectly valid in this scenario and, e.g., the pdf is still useful.  This error should only be thrown when someone attempts to use the relevant algorithms.

Probably having a method on the kde like `ensure_cholesky` that computes and stores the cholesky decomposition should be moved into the methods that require it, rather than [on the constructor](https://github.com/scipy/scipy/blob/v1.14.1/scipy/stats/_kde.py#L571).

### Reproducing Code Example

```python
gaussian_kde(np.array(
  [[1,2],
   [4,6],
   [9,8]]
))
```

### Error message

```shell
ValueError: Number of dimensions is greater than number of samples. This results in a singular data covariance matrix, which cannot be treated using the algorithms implemented in `gaussian_kde`. Note that `gaussian_kde` interprets each *column* of `dataset` to be a point; consider transposing the input to `dataset`
```

### SciPy/NumPy/Python version and system information

```shell
1.14.1 2.1.2 sys.version_info(major=3, minor=11, micro=10, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /opt/_internal/cpython-3.11.9/lib/python3.11/site-packages/scipy_openblas32/include
    lib directory: /opt/_internal/cpython-3.11.9/lib/python3.11/site-packages/scipy_openblas32/lib
    name: scipy-openblas
    openblas configuration: OpenBLAS 0.3.27.dev DYNAMIC_ARCH NO_AFFINITY Zen MAX_THREADS=64
    pc file directory: /project
    version: 0.3.27.dev
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /opt/_internal/cpython-3.11.9/lib/python3.11/site-packages/scipy_openblas32/include
    lib directory: /opt/_internal/cpython-3.11.9/lib/python3.11/site-packages/scipy_openblas32/lib
    name: scipy-openblas
    openblas configuration: OpenBLAS 0.3.27.dev DYNAMIC_ARCH NO_AFFINITY Zen MAX_THREADS=64
    pc file directory: /project
    version: 0.3.27.dev
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.12.0
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.11
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  pythran:
    include directory: ../../tmp/pip-build-env-k6xvzqr5/overlay/lib/python3.11/site-packages/pythran
    version: 0.16.1
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /opt/python/cp311-cp311/bin/python
  version: '3.11'
```

**Comment by nickodell** [comment]
Hi, can you change the title to something sensible?

**Comment by mdhaber** [comment]
> This only affects certain algorithms: a KDE is still perfectly valid in this scenario and, e.g., the pdf is still useful.

The Cholesky decomposition is used (in place of the inverse) in the PDf calculation. How do you propose to calculate the PDF?

**Comment by csbrown** [comment]
Just [add up the .pdfs from the individual kernels](https://en.wikipedia.org/wiki/Kernel_density_estimation#Definition)?

Here is my current naive workaround:

```
class gaussian_kde(object):
    def __init__(self, X, bw_method=1, weights=None):
        self.X = X.T
        self.bw = bw_method
        self.cov = np.diag(np.ones(self.X.shape[1])*np.sqrt(self.bw))
        self.weights = weights
    def pdf(self, X):
        X = X.T
        def pdf_1(x):
            return scipy.stats.multivariate_normal.pdf(X, x, self.cov)
        return np.mean(np.apply_along_axis(pdf_1, 1, self.X), axis=0)
    def int(self, *bounds, step=0.1):
        nd = self.X.shape[1]
        slices = tuple([slice(b[0],b[1],step) for b in bounds])
        grid = np.mgrid[slices].reshape(nd,-1)
        return np.sum(self.pdf(grid)) * step**nd
```

**Comment by mdhaber** [comment]
With this implementation,`multivariate_normal.pdf` will computes a pseudoinverse of `self.cov` using an eigendecomposition separately for each data point. It should also raise an error if the covariance matrix is singular unless you pass `allow_singular=True`. I don't think that will save you time.

That said, I have wanted to replace the implementation to just use `multivariate_normal.pdf`, but that is held up by batch support for linalg functions (e.g. gh-21462).

**Comment by csbrown** [comment]
The covariance matrix in this implementation cannot possibly be singular: it's just a diagonal matrix based on the bandwidth.  This covariance matrix does not depend on the data at all, and so this works when the number of dimensions is less than the number of data points.  I'll remove my complaint about the cholesky decomposition complexity.

**Comment by mdhaber** [comment]
OK, so it sounds like you are really proposing to use a different covariance matrix. Currently, `gaussian_kde` uses the covariance of the data. If you provided fewer data points than the dimensionality of the space, that is going to be singular. Your covariance matrix is a scaled identity matrix. Such an option was discussed in gh-19250 and the associated PR (gh-19297), but feedback was not positive.

**Comment by csbrown** [comment]
One could also just use a flat kernel with mean 0 and covariance identity matrix in nD, and then scale everything with the usual formula K_h((X-x)/h).  It looks like maybe if I freeze my multivariate_normal above in this way, then the cholesky decomposition will only happen once.

**Comment by nickodell** [comment]
> `self.cov = np.diag(np.ones(self.X.shape[1])*np.sqrt(self.bw))`

This assumes that the covariance between any two different variables is zero, correct? I'm not a domain expert here, but from an API consumer's perspective, this seems like a surprising default - i.e. that we would figure out covariance from the data if you pass 10 points and 10 dimensions, but assume it is zero if you pass 9 points and 10 dimensions.

I could support this if the behavior was opt-in, however.

**Comment by csbrown** [comment]
@nickodell I'm not suggesting that y'all replace the current implementation with my very sad one.  :D  I'm just suggesting that there exists a perfectly valid KDE when the number of dimensions is greater than the number of data points, and various things like the pdf and the integral are well defined in this circumstance.  I'm having to use my sad implementation because the scipy implementation breaks for my use case.  I would greatly prefer that the smart scipy `gaussian_kde` supported this, and then I didn't have to create janky workarounds, hence the issue.

**Comment by mdhaber** [comment]
> I'm just suggesting that there exists a perfectly valid KDE when the number of dimensions is greater than the number of data points

This is only true with additional assumptions that are not made in the current implementation. It can be opt-in as @nickodell points out, but it would be surprising to switch between assuming an identity covariance matrix and a covariance matrix estimated from the data based only on the number of points.

Please consider what was proposed in gh-19297. Would that address your need?

**Comment by csbrown** [comment]
Yes, [gh-19297](https://github.com/scipy/scipy/pull/19297) looks more-or-less useful.

I am using the kde on 2D imagery, and I am choosing the bandwidth via cross-validation across many images, so having the bandwidth be chosen by the `gaussian_kde` is not ideal.  I had actually assumed that when I provided a scalar to `gaussian_kde`, then that was the bandwidth, and not just a scaling factor, so I guess there's a lot more issues with using `scipy.stats.gaussian_kde` for me than I had originally realized.  Thanks for the tips.

**Comment by mdhaber** [comment]
This looks like it is a duplicate of gh-21685, now, so I'll close this and refer to it there.
